src/autopas/containers/LinkedCells.h

```diff
@@ -112,31 +112,33 @@ class LinkedCells : public ParticleContainer<Particle, ParticleCell, SoAArraysTy
   }
 
   void updateContainer() override {
-    // @todo optimize
-    std::vector<Particle> invalidParticles;
-// custom reduction with templates not supported
-//#pragma omp parallel reduction(vecMerge: invalidParticles)
-#ifdef AUTOPAS_OPENMP
-#pragma omp parallel
-#endif  // AUTOPAS_OPENMP
-    {
-      std::vector<Particle> myInvalidParticles;
-      for (auto iter = this->begin(); iter.isValid(); ++iter) {
-        myInvalidParticles.push_back(*iter);
-      }
+    auto haloIter = this->getRegionIterator(this->_cellBlock.getHaloBoxMin(), this->_cellBlock.getHaloBoxMax(),
+                                            IteratorBehavior::haloOnly);
+    if (haloIter.isValid()) {
+      utils::ExceptionHandler::exception(
+          "Linked Cells: Halo particles still present when updateContainer was called. First particle found:\n" +
+          haloIter->toString());
+    }
 #ifdef AUTOPAS_OPENMP
-#pragma omp critical
+#pragma omp parallel for
 #endif  // AUTOPAS_OPENMP
-      invalidParticles.insert(invalidParticles.end(), myInvalidParticles.begin(), myInvalidParticles.end());
-    }
-    for (auto &cell : this->_cells) {
-      cell.clear();
-    }
-    for (auto &particle : invalidParticles) {
-      if (inBox(particle.getR(), this->getBoxMin(), this->getBoxMax())) {
-        addParticle(particle);
-      } else {
-        addHaloParticle(particle);
+    for (size_t cellId = 0; cellId < this->getCells().size(); ++cellId) {
+      // if empty
+      if (not this->getCells()[cellId].isNotEmpty()) continue;
+
+      std::array<double, 3> cellLoweCorner, cellUpperCorner;
+      this->getCellBlock().getCellBoundingBox(cellId, cellLoweCorner, cellUpperCorner);
+
+      for (auto &&pIter = this->getCells()[cellId].begin(); pIter.isValid(); ++pIter) {
+        // if not in cell
+        if (notInBox(pIter->getR(), cellLoweCorner, cellUpperCorner)) {
+          // if not in halo
+          if (inBox(pIter->getR(), this->getBoxMin(), this->getBoxMax()))
+            addParticle(*pIter);
+          else
+            addHaloParticle(*pIter);
+          pIter.deleteCurrentParticle();
```

Comment:
nope :P
deleteCurrentParticle, pIter.isValid() (needs Cell::getNumParticles() ) etc. are not thread safe.

Reply:
alternatives:
1. make everything thread safe
2. safe to temporary vector -> add afterwards (with or without parallelization)

Reply:
both alternatives implemented and compared. See text at the top.